### DISCUSSION TITLE
RecommendedBy

### DISCUSSION THREAD

**Discussion by looee1q** [rees46/android-sdk#90] [General]
На данный момент в Android SDK существует два одинаковых класса `RecommendedBy` но в разных пакетах. Один из них [нормальный](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L3), а второй сделан вообще с применением [джава подхода](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/Params.kt#L46). Полагаю, нужно избаваться от лишнего класса. Вопрос как это сделать? Пометить класс, как `Depricated`, потому что им в теории могут пользоваться за пределами SDK?

Еще мне непонятен смысл хранения значения `RecommendedBy` [в локальной переменной](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/data/repositories/recommendation/RecommendationDataSource.kt#L8). Отсутствует концептуальное понимание идеи, которая закладывалась в этот код.

Дело в том, что сохранение значения `RecommendedBy` сейчас осуществляется в двух местах:
1. [При отслеживании любого события мы занулляем переменную](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/features/trackEvent/impl/TrackEventManagerImpl.kt#L32), которая храние текущее значение `RecommendedBy`. Для чего мы это делаем?
2. [При отслеживании просмотра истории](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/stories/StoriesManager.kt#L79) в переменную `RecommendedBy` уходит значение с типом `RecommendedBy.TYPE.STORIES`.

При этом типы `RecommendedBy` не ограничиваются лишь `RecommendedBy.TYPE.STORIES`, а включают в себя и другие [типы](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L7). Однако в SDK отсутствуют места, в которых бы еще происходил вызов [UseCase(а)](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/usecases/recommendation/SetRecommendedByUseCase.kt#L7), ответственного за присвоение нового значения переменной `recommendedBy`. 
При этом и выдача актуального переменной `recommendedBy` происходит всего в одном месте в SDK: [тут](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/features/trackEvent/impl/TrackEventManagerImpl.kt#L29), для формирования запроса при отслеживании действия пользователя.
Соответственно функциональность работы с переменной, которая хранит актуальное значение `RecommendedBy` практичеси отсвутствует (или нереализована полностью).
Сейчас мне не очень понятна какая логика закладывалась в код, связанный с переменной, хранящей значение `RecommendedBy`. А именно: 
- когда должно обновляться ее значение (при каком действии пользователя)? Полагаю эти действия должны соответствовать тем типам, которые перечислены [тут](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L7).

В [документации](https://reference.api.rees46.com/?kotlin#events-tracking) описывается, что параметр `recommendedBy` нужен для отслеживания событий, которые были сделаны пользователем с помощью инструментов платформы. Я не очень пониманию, что это за события такие.

**Comment by TorinAsakura** [comment]
И так…
Я правильно понял, что сейчас у нас есть один и тот же класс с разной логикой?
1. Место первое: При отслеживании событий, где текущая логика перезаписывает значение RecommendedBy. Возникает вопрос: для чего это делается и что делать, чтобы избежать путаницы?
2. Место второе: В истории рекомендаций, где RecommendedBy используется для хранения значений с типами (например, TYPE_STORIES).

То есть:
- Нет ясной структуры, куда и как правильно сохранять данные, связанные с рекомендациями
- Текущая реализация размывает ответственность между местами хранения данных и их обработкой

И предлагаешь ты:
- Организовать логику вокруг значений RecommendedBy таким образом, чтобы SDK чётко разделял, где используется каждое значение
- Возможно, вынести логику обновления значений и взаимодействия с ними в отдельный Use Case
- Убедиться, что сохраняемые данные не теряют смемантику, и не происходят внезапные перезаписи

Верно? Я просто в твоей портянке быстро запутался…

**Reply by looee1q** [reply]
@TorinAsakura 
> Я правильно понял, что сейчас у нас есть один и тот же класс с разной логикой?

> 1. Место первое: При отслеживании событий, где текущая логика перезаписывает значение RecommendedBy. Возникает вопрос: для чего это делается и что делать, чтобы избежать путаницы?
> 2. Место второе: В истории рекомендаций, где RecommendedBy используется для хранения значений с типами (например, TYPE_STORIES).

У нас имеется два одинаковых класса с одинаковой логикой, просто они находятся в разных пакетах: [первый](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L3) и [второй](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/Params.kt#L46). Полагаю, что сначала существовал класс только второй класс (потому что он написан в джава стиле). А потом добавили первый класс, а второй забыли удалить или пометить как `depricated`. При этом сам SDK предоставляет метод для работы и с [первым](https://github.com/rees46/android-sdk/blob/27f4daca19c304932ed849fdf63ad6f0a2473732/personalization-sdk/src/main/kotlin/com/personalization/Params.kt#L151) и со [вторым](https://github.com/rees46/android-sdk/blob/27f4daca19c304932ed849fdf63ad6f0a2473732/personalization-sdk/src/main/kotlin/com/personalization/Params.kt#L147) классом. Эти методы делают одну и ту же работу - добавляют в запрос к серверу одни и те же параметры: `"recommended_by"` и `"recommended_code"`.

При этом в локальную переменную зачем-то сохраняется только первый класс (я не понимаю зачем, поскольку сохраняется либо только `null`, либо только `TYPE_STORIES`; ничего другого в эту переменную сохранится не может). 
Аналогичных методов сохранения для второго класса в SDK нет.

> То есть:

> - Нет ясной структуры, куда и как правильно сохранять данные, связанные с рекомендациями
> - Текущая реализация размывает ответственность между местами хранения данных и их обработкой

Список скорее следующий:
- У меня отсутствует понимание с какой целью мы вообще предоставляем возможность пользователю самому добавлять в запрос к серверу параметры, связанные с `recommended_By`. Согласно [нашей документации](https://reference.api.rees46.com/?kotlin#events-tracking) этот параметр используется для отслеживания событий, которые были сделаны пользователем с помощью инструментов платформы. Исходя из этого описания у меня сложилось представление, что добавление этого параметра мы должны контролировать сами - и добавлять его в какие-то конкретные методы запроса к серверу.
- Непонятно в каких методах запроса к серверу параметр `recommended_By` может использоваться в принципе. В документации параметр `recommended_By` описывается только у двух методов: [отслеживание добавления товара в корзину](https://reference.api.rees46.com/?kotlin#user-added-product-to-cart) и [отслеживание просмотра товара](https://reference.api.rees46.com/?kotlin#user-viewed-a-product). 
- Непонятно почему класс `RecommendedBy` имеет целых [семь типов](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L7). Кажется, что туда намешали всякого лишнего. Например типы ["instant_search" ](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L12) или ["full_search"](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L13). А типы ["bulk"](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L10), ["chain"](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L9) и ["transactional"](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L11) вообще кажется не имеют ничего общего с параметром `"recommended_by"`.

> И предлагаешь ты:
> - Организовать логику вокруг значений RecommendedBy таким образом, чтобы SDK чётко разделял, где используется каждое значение
> - Возможно, вынести логику обновления значений и взаимодействия с ними в отдельный Use Case
> - Убедиться, что сохраняемые данные не теряют смемантику, и не происходят внезапные перезаписи

Прежде чем что-то предлагать относительно работы с параметром `"recommended_by"` внутри SDK:
- хотелось бы сперва понять, как вообще предполагалось использовать параметр `"recommended_by"`: в каких методах запроса он должен иметься и должен ли этот параметр указываться пользователем SDK, или добавление этого параметра должно происходить автоматически в самом SDK.
- относятся ли типы `"chain"`, `"bulk"` и `"transactional"` к параметру `"recommended_by"`. Или это все-же просто отдельные типы?
- аналогичный вопрос относительно типов `"instant_search"` и `"full_search"`. В документации эти параметры характеризуют лишь тип запроса и не имеют никакого отношения к `"recommended_by"`.
- попросить добавить в документацию всю информацию, проясняющую структуру параметра `"recommended_by"` и смысл всех его возможных типов (и перечисление самих типов).

Возможно с частью вопросов может подсказать @noff.

**Reply by noff** [reply]
> У меня отсутствует понимание с какой целью мы вообще предоставляем возможность пользователю самому добавлять в запрос к серверу параметры, связанные с recommended_By. 

На сайте мы берем `recommended_by` из урла при переходе на товар из одного из наших инструментов и дальше передаем его в трекинге событий так, что программисту сайта в принципе ничего не надо с этим делать.

В мобилке нет адресной строки и при переходе на товар из блока рекомендаций или поиска или пуш-уведомления, мы не можем без участия программиста передать этот recommended_by. Поэтому SDK позволяет указать, откуда был клик по товару и передать эту информацию нам.

Важно понять эту логику на прикладном уровне, чтобы сделать все правильно. Как оно работает сейчас (правильно или последний год после всех переделок все работает криво и глючно), сказать не могу. У меня лично нет возможности проверить корректность логики.


> Непонятно в каких методах запроса к серверу параметр recommended_By может использоваться в принципе. 

Любое событие. Более полная документация у нас указана в JS SDK, поэтому вот ссылки:
1. https://reference.api.rees46.com/?javascript#user-viewed-a-product
2. https://reference.api.rees46.com/?javascript#user-added-product-to-cart
3. https://reference.api.rees46.com/?javascript#user-removed-product-from-cart
4. https://reference.api.rees46.com/?javascript#user-purchased-products
5. но это не полный список, надо точно проанализировать места, где еще может идти этот параметр. Я SDK наизусть не знаю.

> Непонятно почему класс RecommendedBy имеет целых [семь типов](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L7). Кажется, что туда намешали всякого лишнего. Например типы ["instant_search" ](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L12)или ["full_search"](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L13). А типы ["bulk"](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L10), ["chain"](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L9) и ["transactional"](https://github.com/rees46/android-sdk/blob/fcc159117d985510f27e11af9a6cf92413d8158c/personalization-sdk/src/main/kotlin/com/personalization/sdk/domain/models/RecommendedBy.kt#L11) вообще кажется не имеют ничего общего с параметром "recommended_by".

Тут важно понимать следующее:

1. Если юзер переходит на товар, то это скорее всего один из инструментов: popup, instant_search, full_search, dynamic (старый нейминг, это товарные рекомендации), `stories` и что-то еще наверно.
2. Если юзер переходит в приложение из мобильного пуша или из СМС по диплинку, то к себе мы приписываем не конкретный товар, а всю сессию. Т.е. юзер получил мобильный пуш, ткнул в него, добавил товар в корзину и купил - мы весь заказ учтем как заказ от рассылки. Типы рассылок: bulk (массовые), chain (триггерные) и transactional (транзакционные).

Опять же, за адекватность реализации текущей версии android SDK гарантировать не могу.


Вообще лучше голосом быстро пройти с просмотром кода, а то я все это в слепую пишу, в код не заглядываю.

**Reply by TorinAsakura** [reply]
@looee1q мы можем не депрекейтить старый класс который в параметрах живёт (чо он там делает - конечно загадка), так как у нас семвер - мы просто пишем об этом в разделе брикин чейнджс и на этом наши полномочия всё

По поводу того мусора что живёт в параметрах - это точно нужно будет выпиливать, так как оно точно не соответствует той логике, что сейчас описал Миша (да и нашей доке в целом)

**Reply by looee1q** [reply]
@TorinAsakura
> мы можем не депрекейтить старый класс который в параметрах живёт

Значит его можно просто удалить?

> мы просто пишем об этом в разделе брикин чейнджс

Что это за раздел такой? И кто туда должен писать?

**Reply by TorinAsakura** [reply]
> @TorinAsakura
> 
> > мы можем не депрекейтить старый класс который в параметрах живёт
> 
> Значит его можно просто удалить?

- верно

> > мы просто пишем об этом в разделе брикин чейнджс
> 
> Что это за раздел такой? И кто туда должен писать?

- в ченйджлоге @Nelfimov подскажет

**Reply by Nelfimov** [reply]
> > > мы просто пишем об этом в разделе брикин чейнджс
> > 
> > 
> > Что это за раздел такой? И кто туда должен писать?
> 
> * в ченйджлоге @Nelfimov подскажет

@looee1q для наших пакетов автоматом генерируется чейнджлог по мержу в мастер. Коммиты парсятся и фомируется чейнджлог. Т.к. мы следуем конвенции коммитов по ангуляру, то чейнджлог создается красивым:

https://github.com/rees46/web/blob/stage/packages/segmentator/CHANGELOG.md

В веб разработке используем https://www.npmjs.com/package/conventional-changelog-cli

Для мобил придется искать аналог

**Reply by looee1q** [reply]
@Nelfimov 
> для наших пакетов автоматом генерируется чейнджлог по мержу в мастер.

Что такое пакет в данном контексте? Это какая-то совокупность ПР или коммитов?

**Reply by Nelfimov** [reply]
мы работаем с монорепозиториями - организация репозитория в котором хранится несколько пакетов/приложений.

У тебя монорепозиторий, поэтому достаточно одного чейнджлога.

**Reply by looee1q** [reply]
@TorinAsakura В Android SDK нужно тоже ввести файл аналогичный [changelog](https://github.com/rees46/web/blob/stage/packages/segmentator/CHANGELOG.md) из веба?

**Reply by TorinAsakura** [reply]
> @TorinAsakura В Android SDK нужно тоже ввести файл аналогичный [changelog](https://github.com/rees46/web/blob/stage/packages/segmentator/CHANGELOG.md) из веба?

конечно, у нас ведь должны быть релизы, мы ведь публичный сдк ведём
